### PR TITLE
fix `NamedArray.imag` and `NamedArray.real` typing info

### PR DESCRIPTION
**Pull request by andersy005**
- [ ] #8365 follow up
- [ ] Tests added

### CODE REVIEW

**Comment by andersy005** [comment]
> I wonder if it's better to use the array_api functions?

last night i tried, and it appears it won't work until there's wide support of the array API in upstream dependencies (e.g. pint). 

```python
In [14]: import pint, xarray as xr, numpy as np

In [15]: unit_registry = pint.UnitRegistry(force_ndarray_like=False)

In [16]: Quantity = unit_registry.Quantity

In [17]: dtype = float

In [18]: array = (
    ...:             np.arange(5 * 10).astype(dtype)
    ...:             + 1j * np.linspace(-1, 0, 5 * 10).astype(dtype)
    ...:         ).reshape(5, 10) * unit_registry.s

In [23]: np.real(array)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[23], line 1
----> 1 np.real(array)

File <__array_function__ internals>:200, in real(*args, **kwargs)

TypeError: no implementation found for 'numpy.real' on types that implement __array_function__: [<class 'pint.Quantity'>]
```

curious to hear if there is a consistent way of using the array api functions as they are today without needing to special case upstream libraries with limited or without array api compliance.

File: xarray/core/variable.py [hunk 1]
```diff
@@ -2365,6 +2371,32 @@ def notnull(self, keep_attrs: bool | None = None):
             keep_attrs=keep_attrs,
         )
 
+    @property
+    def imag(
+        self: Variable[_ShapeType, np.dtype[_SupportsImag[_ScalarType]]]
+    ) -> Variable[_ShapeType, np.dtype[_ScalarType]]:
```
Comment (andersy005) [inline xarray/core/variable.py:2377]: mypy isn't happy :( 

```
xarray/core/variable.py: note: In member "imag" of class "Variable":
xarray/core/variable.py:2376: error: "Variable" expects no type arguments, but 2 given  [type-arg]
xarray/core/variable.py:2377: error: "Variable" expects no type arguments, but 2 given  [type-arg]
xarray/core/variable.py: note: In member "real" of class "Variable":
xarray/core/variable.py:2389: error: "Variable" expects no type arguments, but 2 given  [type-arg]
xarray/core/variable.py:2390: error: "Variable" expects no type arguments, but 2 given  [type-arg]
xarray/core/alignment.py: note: In function "align":
```

Comment (Illviljan) [reply xarray/core/variable.py:2377]: Yeah, Variable doesn't have the shape and dtype generics (`Generic[_ShapeType_co, _DType_co]`) currently. 

Variable doesn't have the strict mypy requirement so I think you can get away with simply:
```python
def imag(self) -> Variable:
    ...
```

**Comment by andersy005** [comment]
thank you for all your help, @Illviljan! i believe this is ready for a last round of reviews.

File: xarray/namedarray/_typing.py [hunk 2]
```diff
@@ -138,6 +138,14 @@ def __array_function__(
     ) -> Any:
         ...
 
+    @property
+    def imag(self) -> _arrayfunction[_ShapeType_co, Any]:
+        ...
+
+    @property
+    def real(self) -> _arrayfunction[_ShapeType_co, Any]:
+        ...
```
Comment (Illviljan) [inline xarray/namedarray/_typing.py:147]: Not sure why the numpy strategy of replacing self with a new type doesn't work here.
Oh well, NamedArray has the correct dtype so it's good enough for me.

**Review by Illviljan — APPROVED** [review]
Well done, looking good to me!

**Pull request merged**